Add redgtech integration

Jonhsady opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37217

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Jonhsady committed:
add redgtech integration

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
Merge branch 'dev' into dev

zweckj reviewed (changes requested):
Hi, thanks for contributing. However, there's a couple of things wrong with this PR

- it's missing a meaningful description
- it's adding more than one platform, which [the bot already told you](https://github.com/home-assistant/core/pull/136947#pullrequestreview-2584830151)
- it doesn't use a library on PyPi for abstracting the actual API communication

please check the [Development Checklist](https://developers.home-assistant.io/docs/development_checklist)

Thanks 👍

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

zweckj reviewed (changes requested):
I don't want to be rude, but please, read the development guidelines.
- you still have more than one platform (light & switch)
- your library code is not publicly avaiable on a repository (not PyPi), OSI license, automated & tagged releases
- you're quality scale is way off (e.g. you DO have stateful entities)
- only by glancing at your code I can tell you didn't really read the quality scale criteria (e.g. `runtime-data`)

Jonhsady commented:
> I don't want to be rude, but please, read the development guidelines.
> 
> * you still have more than one platform (light & switch)
> * your library code is not publicly avaiable on a repository (not PyPi), OSI license, automated & tagged releases
> * you're quality scale is way off (e.g. you DO have stateful entities)
> * only by glancing at your code I can tell you didn't really read the quality scale criteria (e.g. `runtime-data`)

Jonhsady commented:
> I don't want to be rude, but please, read the development guidelines.
> 
> * you still have more than one platform (light & switch)
> * your library code is not publicly avaiable on a repository (not PyPi), OSI license, automated & tagged releases
> * you're quality scale is way off (e.g. you DO have stateful entities)
> * only by glancing at your code I can tell you didn't really read the quality scale criteria (e.g. `runtime-data`)


Calm down senior, you seem nervous to me, it's just that it's very confusing to do the integration, peace! I will make the necessary changes and if there is a mistake, please don't be rude

Jonhsady committed:
remove plataform light

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

Jonhsady commented:
@zweckj 
Thank you very much for the feedback, you are really right, I made the changes and I hope it works this time

luan-nvg committed:
Merge branch 'dev' into dev

Jonhsady committed:
adjustements manifest

Jonhsady committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+            
+            try:
+                async with aiohttp.ClientSession() as session:
+                    async with session.post(
+                        f'{API_URL}/home_assistant/login',
+                        json={'email': email, 'password': password}
```
we don't allow naked HTTP requests in core, those are to be abstracted away in a lib

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,69 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Redgtech from a config entry."""
+    _LOGGER.debug("Setting up Redgtech entry: %s", entry.entry_id)
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        "config": entry.data,
+        "entities": []
+    }
```
use `entry.runtime_data` instead (check the quality scale)

zweckj commented on homeassistant/components/redgtech/services.yaml:
no services

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,71 @@
+        else:
+            _LOGGER.error("Failed to set state for %s", self._name)
+
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            "endpoint_id": self._endpoint_id,
+            "description": self._description,
+            "manufacturer": self._manufacturer,
```
those are things for a `device`

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,71 @@
+    else:
+        _LOGGER.error("No access token available")
+
+class RedgtechSwitch(SwitchEntity):
```
currently that switch would never update, and therefore not get informed about changes outside of HA

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,71 @@
+    access_token = config_entry.data.get("access_token")
+    if access_token:
+        api = RedgtechAPI(access_token)
+        try:
+            data = await api.get_data()
+            entities = []
+            for item in data.get("boards", []):
+                categories = item.get("displayCategories", "")
+                if "SWITCH" in categories:
+                    entities.append(RedgtechSwitch(item, api))
```
I think it makes sense to implement a `DataUpdateCoordinator` for you

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
remove coments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
please add a custom type for your entry

zweckj replied:
@luan-nvg I wanted to give this another pass, but this is a perfect example, this conversation is resolved and the [runtime-data](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/runtime-data/) item in the quality scale is marked as `done`, but it's not addressed and if I look at the mentioned documentation, there's a pretty clear description what's to do, leading me to believe this documentation has not been read properly.

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+
+    VERSION = 1
+
+    async def async_step_user(self, user_input=None):
```
types

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+class RedgtechConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config Flow for Redgtech integration."""
+
+    VERSION = 1
```
is default
```suggestion
```

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+        errors = {}
+        
+        if user_input is not None:
+            email = user_input.get("email")
+            password = user_input.get("password")
```
don't get when you don't have to. Those are required, so they will be there

zweckj replied:
```suggestion
            email = user_input.get(CONF_EMAIL)
            password = user_input.get(CONF_PASSWORD)
```
we use consts for those things `from homeassistant.const`

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+            try:
+                access_token = await api.login(email, password)
+                if access_token:
+                    _LOGGER.info("Login successful")
```
info logging is reserverd for the system, use debug

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial user step for login."""
+        errors = {}
```
type

zweckj replied:
still missing a correct type

zweckj replied:
this is still missing

Jonhsady replied:
async def async_step_user(self, user_input: Optional[Dict[str, Any]] = None) -> FlowResult:

I'm sorry, but doesn't this solve the issue of type?

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+
+                    return self.async_create_entry(
+                        title="Redgtech",
+                        data={"access_token": access_token}
```
```suggestion
                        data={CONF_ACCESS_TOKEN: access_token}
```

zweckj commented on homeassistant/components/redgtech/services.yaml:
drop the file

zweckj commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,67 @@
+            "user": {
+                "title": "User Configuration",
+                "description": "Please enter your email address.",
+                "data": {
```
missing data description, missing errors for the config flow, lots of extra unused keys

zweckj commented on homeassistant/components/redgtech/manifest.json:
```diff
@@ -0,0 +1,12 @@
+    "logo": "/brands/redgtech/logo.png",
+    "integration_type": "service",
+    "config_flow": true,
+    "quality_scale": "silver",
```
```suggestion
    "quality_scale": "bronze",
```
you won't reach anything more with this PR

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+    comment: only entity actions
+  appropriate-polling:
+    status: exempt
+    comment: the integration does not poll
```
please re-review those a lot of those are not true

Jonhsady replied:
I'm sorry Zwe, but do you suggest gold?

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  parallel-updates: done
+  reauthentication-flow:
+    status: exempt
+    comment: the integration currently does not implement authenticated requests
```
you have an access authentication token?

zweckj commented on homeassistant/components/redgtech/manifest.json:
```diff
@@ -0,0 +1,12 @@
+    "codeowners": [],
+    "documentation": "https://www.home-assistant.io/integrations/redgtech",
+    "iot_class": "cloud_polling",
+    "logo": "/brands/redgtech/logo.png",
```
```suggestion
```
will come from the brands repo

zweckj commented on homeassistant/components/redgtech/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "domain": "redgtech",
+    "name": "Redgtech",
+    "codeowners": [],
```
is required

zweckj commented on homeassistant/components/redgtech/manifest.json:
```diff
@@ -0,0 +1,12 @@
+    "integration_type": "service",
+    "config_flow": true,
+    "quality_scale": "silver",
+    "requirements": ["redgtech-api==0.1.14"]
```
where is that code of your library? It should be in a publicly available repo, with CI and tagged releases and an OSI approved license

Jonhsady replied:
https://pypi.org/project/redgtech-api/0.1.16/

Jonhsady committed:
adjustments for colaborator

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,115 @@
+        else:
+            _LOGGER.error("Failed to set state for %s", self._name)
+
+    async def async_update(self):
+        """Fetch new state data for the switch."""
+        _LOGGER.debug("Updating switch state: %s", self._name)
+        await self.coordinator.async_request_refresh()
+        data = self.coordinator.data
+        if data:
+            for item in data.get("boards", []):
+                if item.get("endpointId") == self._endpoint_id:
+                    self._state = item.get("value", False)
+                    self.async_write_ha_state()
```
coordinator will handle the update

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,115 @@
+                    self._state = item.get("value", False)
+                    self.async_write_ha_state()
+
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            "endpoint_id": self._endpoint_id,
+        }
```
we don't like extra_state_attributes anymore and I don't see the value this adds
```suggestion
```

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,115 @@
+class RedgtechSwitch(SwitchEntity):
+    """Representation of a Redgtech switch."""
+
+    def __init__(self, coordinator, data, api):
```
please, actually read the items required by the quality scale.

This is missing a name and `_attr_has_entity_name = True`, which you'd know if you actually read those. It's also missing a unique id, all things explained in the quality scale.

Jonhsady committed:
add coordinator file

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,47 @@
+    api = RedgtechAPI(access_token)
+    coordinator = RedgtechDataUpdateCoordinator(hass, api)
+    
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except Exception as e:
+        _LOGGER.error("Error fetching data from API: %s", e)
+        return False
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```
coordinator handles errors

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,47 @@
+        _LOGGER.error("Error fetching data from API: %s", e)
+        return False
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = RedgtechEntryData(
+        config=entry.data,
+        coordinator=coordinator
+    )
```
use entry.runtime_data, don't store the entry.data in the runtime data, the coordinator is enough

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,52 @@
+        errors = {}
+        
+        if user_input is not None:
+            email = user_input.get(CONF_EMAIL)
```
you are still getting instead if direct accessing []

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,52 @@
+                if access_token:
+                    _LOGGER.debug("Login successful, token received.")
+
+                    existing_entries = self._async_current_entries()
+                    for entry in existing_entries:
+                        if entry.data.get(CONF_ACCESS_TOKEN) == access_token:
+                            return self.async_abort(reason="already_configured")
```
Would an access token for the same email and password actually be the same for two entries? Rather use `abort_entries_match()` CONF_EMAIL

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+        try:
+            _LOGGER.debug("Fetching data from Redgtech API")
+            data = await self.api.get_data()
+            entities = []
```
missing a type

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+class RedgtechDataUpdateCoordinator(DataUpdateCoordinator):
```
still missing a type

zweckj replied:
and this is also still missing the return type. Sorry, but I'm losing my patience here. Go over the previous comments again and try to solve them (all of them).

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+class RedgtechDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the API."""
+
+    def __init__(self, hass, api: RedgtechAPI):
```
also pass the entry as param

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=1),
```
config_entry
```suggestion
            update_interval=timedelta(seconds=1),
            config_entry=config_entry,
```

zweckj replied:
that's too often, maximum acceptable is 5 seconds

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+
+    def __init__(self, hass, api: RedgtechAPI):
+        """Initialize."""
+        self.api = api
```
you can build it here, so you don't have to pass it as param

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
```
still missing typing

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,72 @@
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the switch platform."""
+    entry_data = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator = entry_data.coordinator
```
```suggestion
    coordinator = entry.runtime_data
```

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,72 @@
+    entry_data = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator = entry_data.coordinator
+
+    entities = []
+    if coordinator.data:
+        existing_entities = hass.data.get(DOMAIN, {}).get("entities", [])
+        for item in coordinator.data:
+            entity_id = item["id"]
+            if entity_id not in existing_entities:
+                if item["type"] == "switch":
+                    entities.append(RedgtechSwitch(coordinator, item))
+                    existing_entities.append(entity_id)
+
+        hass.data.setdefault(DOMAIN, {})["entities"] = existing_entities
```
there's no need to check for any existing entities, this is only called once, just add the ones you need

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,72 @@
+            self._state = state
+            self.async_write_ha_state()
+            _LOGGER.debug("State of %s set to %s", self._name, state)
+            await self.coordinator.async_request_refresh()
```
why are you doing another update, you already updated the value manually

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,72 @@
+            _LOGGER.debug("State of %s set to %s", self._name, state)
+            await self.coordinator.async_request_refresh()
+        else:
+            _LOGGER.error("Failed to set state for %s", self._name)
```
```suggestion
            raise HomeAssistantError(...)
```

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+            update_interval=timedelta(seconds=1),
+        )
+
+    async def _async_update_data(self) -> List[dict]:
```
please type better. Add a dataclass for your entities if you have nothing in the lib. And I'd suggest another name for those to not mix them with the HA entities

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+    async def _async_update_data(self) -> List[dict]:
+        """Fetch data from API."""
+        try:
+            _LOGGER.debug("Fetching data from Redgtech API")
```
please, only code in try that can actually fail

zweckj reviewed (changes requested):
you did only process half the feedback I gave you last time, please be more diligent. Ask if you don't understand something before setting back to "Ready for review"

Jonhsady committed:
adding types

Jonhsady committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,35 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = RedgtechEntryData(
+        coordinator=coordinator
+    )
```
I already told you to to remove that
```suggestion
    entry.runtime_data = RedgtechEntryData(
        coordinator=coordinator
    )
```

Jonhsady committed:
adjustements

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

NoRi2909 commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,27 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "User Configuration",
```
Please change to "User configuration" (sentence-casing in HA)

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

luan-nvg committed:
Merge branch 'dev' into dev

Jonhsady committed:
Merge branch 'dev' into dev

luan-nvg commented:
Hi @zweckj , good morning, how are you? I know I'm commenting on a serious infraction by tagging you, but we're worried. We're a sturtup that's just starting out and John is just an intern trying to survive in technology. I understand your frustration, but please be patient with him. The home assistant documentation is indeed complete, but complex for us, which is why there's been a lot of back and forth. John has been staying up all night trying to get HA to work, so don't be so upset with him. Okay? We're learning English and learning to understand HA, which is painful for non-native speakers. He sent the changes 3 weeks ago, and now we don't know if we should wait or if we have some other pending issue with you. Please have compassion for us. We're here to bring more technology to our country.

luan-nvg committed:
Merge branch 'dev' into dev

zweckj commented:
> [...] but we're worried. 

No reason for that.

> [...] but please be patient with him. [...] so don't be so upset with him. 

I am and I am not upset. But if I need request changes multiple times and you mark them resolved without addressing them, you have to understand this is arduous. I am doing this on my free time (for free), so if I feel like my time is better invested somewhere else I will do so.

> [...] now we don't know if we should wait

Yes. I'll look at this again when I have some time (or somebody else might). There is a queue of 500 PRs and new integrations tend to take longer.

zweckj commented:
You don't have to keep merging dev into this PR.

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,35 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+@dataclass
+class RedgtechEntryData:
+    coordinator: RedgtechDataUpdateCoordinator
```
```suggestion
type RedgtechConfigEntry = ConfigEntry[RedgtechDataUpdateCoordinator]
```

zweckj replied:
there will definitely be more things. But as I tried to explain already: If we have to do this back and forth for every single comment multiple times it's exhausting and to be honest doesn't make me want to spend my free time looking at this.

Jonhsady committed:
adjustments quality_scale

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
adjustments requirements_all

Jonhsady committed:
adjustments manifest

luan-nvg committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+        super().__init__(coordinator)
+        self.coordinator = coordinator
+        self.device = device
+        self.api: Any = coordinator.api
```
```suggestion
        self.api = coordinator.api
```

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+        self.device = device
+        self.api: Any = coordinator.api
+        self._state: bool = device.state == STATE_ON
+        self._name: str = device.name
+        self._endpoint_id: str = device.id
+        self._attr_unique_id: str = f"redgtech_{self._endpoint_id}"
```
```suggestion
        self._name = device.name
        self._endpoint_id = device.id
        self._attr_unique_id = f"redgtech_{self._endpoint_id}"
```

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+        self._endpoint_id: str = device.id
+        self._attr_unique_id: str = f"redgtech_{self._endpoint_id}"
+
+    @property
+    def name(self) -> str:
+        """Return the name of the switch."""
+        return self.device.name
```
```suggestion
```
is the switch the main feature of the device? If yes, set `_attr_entity_name` to `None`, so it'll follow the device name (which you have to define as `_attr_device_info`

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+        """Turn the switch on."""
+        try:
+            await self.coordinator.set_device_state(self.device.id, True)
+            await self.coordinator.async_request_refresh()
```
why are you refreshing?

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+            await self.coordinator.async_request_refresh()
+            self._state = self.device.state == STATE_ON
+            self.async_write_ha_state()
+        except Exception as e:
+            raise HomeAssistantError(f"Failed to update switch {self.device.name}: {e}")
```
don't catch bare exceptions

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,81 @@
+        except Exception as e:
+            raise HomeAssistantError(f"Failed to turn off switch {self.device.name}: {e}")
+
+    async def async_update(self) -> None:
+        """Fetch new state data for the switch."""
+        try:
+            await self.coordinator.async_request_refresh()
+            self._state = self.device.state == STATE_ON
+            self.async_write_ha_state()
+        except Exception as e:
```
you have a coordinator, your device will update thorugh that
```suggestion
```

Jonhsady replied:
Below you asked me to delete the coordinator update and keep it on the switch, so is this code still kept here?

zweckj replied:
No the update stays on the coordinator only. The actual switching code moves to the switch

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+
+        return devices
+
+    async def set_device_state(self, device_id: str, state: str):
+        """Set the state of a device."""
+        success = await self.api.set_switch_state(device_id, state, self.access_token)
+        if success:
+            await self.async_request_refresh()
+        else:
+            raise HomeAssistantError(f"Failed to set state for {device_id}")
```
```suggestion
```
you currently have the code in here basically twice, let's keep it in the switch only where it belongs

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
please re-evaluate the quality scale, try to understand the documentation

Jonhsady commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+                access_token = await coordinator.login(email, password)
+                if not access_token:
+                    raise InvalidAuth
+            except InvalidAuth:
+                _LOGGER.error("Login failed: Invalid authentication")
+                errors["base"] = "invalid_auth"
+            except CannotConnect:
```
Does a generic error solve the problem or are custom API errors ideal?

zweckj replied:
a generic `except Exception` would work here (that is the only place where it's allowed), but specific exceptions from the library would of course be better, as this allows you to give the correct (translated) problem to the user.

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=5),
```
Actually I don't think so, I think I'll apply updates on demand

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+    async def login(self, email: str, password: str) -> str:
+        """Login to the Redgtech API and return the access token."""
+        try:
+            access_token = await self.api.login(email, password)
+            self.access_token = access_token
```
no

zweckj replied:
so you have tokens in your system that are valid indefinitely? That seems strange. What if someone get's access to my password, gets a token. Then that token would still be valid even when I change my password

Jonhsady replied:
Oh sorry, I misunderstood, the token changes if you change your password, etc.

zweckj replied:
But you don't have to renew your token otherwise?

Jonhsady replied:
sim, então ao renovar o token tenho q redirecionar pra tela de login do HA? como posso fazer? é para chamar alguma validacao nas rotas que estao sendo executadas no HA (exceto login)?

Jonhsady replied:
yes, so when renewing the token do I have to redirect to the HA login screen? How can I do this? Is it to call some validation on the routes that are being performed in HA (except login)?

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+        except Exception as e:
+            raise UpdateFailed(f"Error fetching data: {e}")
+
+        devices: List[RedgtechDevice] = []
```
from typing import List, Optional

because with list and not with List? with list there is no type.

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

Jonhsady committed:
Merge branch 'dev' into dev

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,85 @@
+
+class RedgtechSwitch(CoordinatorEntity[RedgtechDataUpdateCoordinator], SwitchEntity):
+    """Representation of a Redgtech switch."""
+
```
_attr_has_entity_name = True
_attr_name = None

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,85 @@
+            model="Switch Model"
+        )
+    
+    @property
+    def name(self) -> str:
+        """Return the name of the switch."""
+        return self.device.name
```
```suggestion
    @property
    def name(self) -> str:
        """Return the name of the switch."""
        return self.device.name
```
you just ignored my previous comment once again...

Jonhsady replied:
It's because the channel names didn't turn out the way I wanted.

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,85 @@
+        """Return true if the switch is on."""
+        return self.device.state == STATE_ON
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the switch on."""
+        try:
+            await self.coordinator.set_device_state(self.device.id, True)
+            self.device.state = STATE_ON
+            self.async_write_ha_state()
+        except ConfigEntryError as e:
+            _LOGGER.error("Failed to turn on switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Failed to turn on switch {self.device.name}: {e}") from e
+        except RedgtechConnectionError as e:
+            _LOGGER.error("Network error while turning on switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Network error while turning on switch {self.device.name}: {e}") from e
+        except Exception as e:
+            _LOGGER.error("Unexpected error while turning on switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Unexpected error while turning on switch {self.device.name}: {e}") from e
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the switch off."""
+        try:
+            await self.coordinator.set_device_state(self.device.id, False)
+            self.device.state = STATE_OFF
+            self.async_write_ha_state()
+        except ConfigEntryError as e:
+            _LOGGER.error("Failed to turn off switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Failed to turn off switch {self.device.name}: {e}") from e
+        except RedgtechConnectionError as e:
+            _LOGGER.error("Network error while turning off switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Network error while turning off switch {self.device.name}: {e}") from e
+        except Exception as e:
+            _LOGGER.error("Unexpected error while turning off switch %s: %s", self.device.name, e)
+            raise HomeAssistantError(f"Unexpected error while turning off switch {self.device.name}: {e}") from e
```
implement `async_toggle` instead. Move the code from the coordinator here. Don't log errors x-times.
```suggestion
```

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,114 @@
+
+        return devices
+
+    async def set_device_state(self, device_id: str, state: bool) -> None:
+        """Set the state of a device."""
+        _LOGGER.debug("Setting device state: %s to %s", device_id, state)
+        try:
+            await self.api.set_switch_state(device_id, state, self.access_token)
+        except RedgtechAuthError:
+            _LOGGER.warning("Access token expired, attempting to renew")
+            await self.renew_token(self.config_entry.data["email"], self.config_entry.data["password"])
+            await self.api.set_switch_state(device_id, state, self.access_token)
+        except RedgtechConnectionError as e:
+            _LOGGER.error("Connection error while setting device state: %s", e)
+            raise ConfigEntryError("Failed to set device state") from e
+        except Exception as e:
+            _LOGGER.error("Unexpected error setting device state: %s", e)
+            raise HomeAssistantError("Unexpected error setting device state") from e
+        _LOGGER.debug("Device state set successfully")
+        await self._async_update_data()
```
just move all that code into `async_toggle` in the switch
```suggestion
```

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,77 @@
+
+class RedgtechDataUpdateCoordinator(DataUpdateCoordinator [List[RedgtechDevice]]):
+    """Coordinator to manage fetching data from the Redgtech API."""
+
```
"Hi @zweck, to resolve the circular import issue with RedgtechConfigEntry, would it be acceptable to move its definition to a separate models.py file within the integration? This would centralize shared types and avoid circular dependencies between __init__.py and [coordinator.py](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html). Let me know if this aligns with the project's structure and guidelines. Thanks!"

zweckj replied:
should not be necessary, it should be fine to only have it in coordinator.py (you might need `__future__.annotations`), check the other core integrations

Jonhsady committed:
Merge branch 'dev' into dev

Jonhsady committed:
adjustments

Jonhsady committed:
Merge branch 'dev' of https://github.com/redgtech-automacao/core into dev

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+from redgtech_api.api import RedgtechAPI, RedgtechAuthError, RedgtechConnectionError
+from .const import DOMAIN
+
+if TYPE_CHECKING:
+    from . import RedgtechConfigEntry
```
it's ok, but if you just move the type definition here, you wouldn't need this

Jonhsady replied:
class RedgtechConfigEntry(ConfigEntry):
    """Custom type for Redgtech config entries."""
    runtime_data: RedgtechDataUpdateCoordinator  
    
    
    resolve?

Jonhsady replied:
resolve?

class RedgtechConfigEntry(ConfigEntry):
    """Custom type for Redgtech config entries."""
    runtime_data: RedgtechDataUpdateCoordinator

zweckj replied:
if they are both inside the `coordinator` file and you have `__future__.annotations` this should just work

zweckj replied:
see e.g. https://github.com/home-assistant/core/blob/dev/homeassistant/components/tedee/coordinator.py#L36-L42

Jonhsady replied:
Hmm, it's in my init. Should I move it to COORDINATOR?

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+    from . import RedgtechConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.debug("Coordinator for Redgtech is being initialized.")
```
```suggestion
```
no logs outside of methods

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+        """Return true if the switch is on."""
+        return self.device.state == STATE_ON
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the switch on."""
+        await self._async_set_state(True)
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the switch off."""
+        await self._async_set_state(False)
```
```suggestion
```
you have toggle

Jonhsady replied:
resolve?

  async def async_turn_on(self, **kwargs: Any) -> None:
        """Turn the switch on."""
        await self.async_toggle()

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn the switch off."""
        await self.async_toggle()

Jonhsady replied:
resolve?


  async def async_turn_on(self, **kwargs: Any) -> None:
        """Turn the switch on."""
        await self.async_toggle()

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn the switch off."""
        await self.async_toggle()

zweckj replied:
no you don't need it at all

Jonhsady replied:
Removing these methods and keeping only async_toggle() breaks the integration with the interface, because ToggleEntity defines that the entity must know how to handle these operations explicitly.

zweckj replied:
ah misremembered, sorry. What you said then:

```python
async def async_turn_on(self, **kwargs: Any) -> None:
"""Turn the switch on."""
await self.async_toggle()
async def async_turn_off(self, **kwargs: Any) -> None:
"""Turn the switch off."""
await self.async_toggle()
```

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+        new_state = not self.is_on
+        await self._async_set_state(new_state)
+
+    async def _async_set_state(self, state: bool) -> None:
+        """Set the state of the switch."""
+        try:
+            await self.coordinator.api.set_switch_state(self.device.id, state, self.coordinator.access_token)
+            self.device.state = STATE_ON if state else STATE_OFF
+            self.async_write_ha_state()
+        except RedgtechAuthError:
+            await self.coordinator.renew_token()
+            await self.coordinator.api.set_switch_state(self.device.id, state, self.coordinator.access_token)
+            self.device.state = STATE_ON if state else STATE_OFF
+            self.async_write_ha_state()
+        except RedgtechConnectionError as e:
+            raise ConfigEntryError("Failed to set switch state due to connection error") from e
+        except Exception as e:
+            raise HomeAssistantError("Unexpected error while setting switch state") from e
```
move this instide toggle

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+        """Set the state of the switch."""
+        try:
+            await self.coordinator.api.set_switch_state(self.device.id, state, self.coordinator.access_token)
+            self.device.state = STATE_ON if state else STATE_OFF
```
move this as the first item in the method so you don't need this twice

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+        try:
+            await self.coordinator.api.set_switch_state(self.device.id, state, self.coordinator.access_token)
+            self.device.state = STATE_ON if state else STATE_OFF
+            self.async_write_ha_state()
```
move this as the last thing in the method

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+    coordinator: RedgtechDataUpdateCoordinator = config_entry.runtime_data
+    async_add_entities(RedgtechSwitch(coordinator, device) for device in coordinator.data)
+
+class RedgtechSwitch(ToggleEntity, RestoreEntity):
```
why is this a `RestoreEntity`?

zweckj replied:
the question was rather: why would the state not be available, you are refreshing all data before the integration loads?

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+    hass: HomeAssistant, config_entry: RedgtechConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the switch platform."""
+    coordinator: RedgtechDataUpdateCoordinator = config_entry.runtime_data
```
```suggestion
    coordinator = config_entry.runtime_data
```
type is inferred

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+from homeassistant.helpers.entity import DeviceInfo, ToggleEntity
+from homeassistant.helpers.restore_state import RestoreEntity
+
+if TYPE_CHECKING:
+    from . import RedgtechConfigEntry
```
```suggestion
```

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+        """Renew the access token."""
+        _LOGGER.debug("Renewing access token")
+        try:
+            new_access_token = await self.api.login(
```
so either you don't store the access token in the entry at all, or you need to update the entry here. You said the token won't expire normally, right?

Jonhsady replied:
Yes, the token does not expire, only if the password is changed, deleted or something like that.

zweckj replied:
I'd probably say don't store the access token in the entry then (don't store `CONF_ACCESS_TOKEN`), but get it at each start, since access token and password are tied together in their lifecycle anyways

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+        except RedgtechAuthError as e:
+            raise ConfigEntryError("Authentication error while renewing access token") from e
+        except RedgtechConnectionError as e:
+            raise ConfigEntryError("Connection error while renewing access token") from e
```
```suggestion
            raise ConfigEntryNotReady("Connection error while renewing access token") from e
```
this seems like an intermittent error

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+            await self.coordinator.api.set_switch_state(self.device.id, state, self.coordinator.access_token)
+            self.device.state = STATE_ON if state else STATE_OFF
+            self.async_write_ha_state()
+        except RedgtechConnectionError as e:
+            raise ConfigEntryError("Failed to set switch state due to connection error") from e
```
we should only raise `HomeAssistantErrors` in actions

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,98 @@
+from redgtech_api.api import RedgtechAuthError, RedgtechConnectionError
+
+
+@pytest.mark.asyncio
```
```suggestion
```
remove that everywhere

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,98 @@
+        CONF_PASSWORD: "123456"
+    }
+
+    with patch("homeassistant.components.redgtech.config_flow.RedgtechAPI.login", return_value="fake_token"):
+        result = await mock_flow.async_step_user(user_input)
```
use a fixture instead

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,98 @@
+    with patch("homeassistant.components.redgtech.config_flow.RedgtechAPI.login", return_value="fake_token"):
+        result = await mock_flow.async_step_user(user_input)
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```
everywhere, we compare enums with `is`

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,98 @@
+@pytest.mark.asyncio
+async def test_user_step_creates_entry():
+    """Test if the configuration flow creates an entry correctly."""
+    mock_hass = MagicMock()
+    mock_hass.config_entries.async_entries = MagicMock(return_value=[])
+
+    mock_flow = RedgtechConfigFlow()
+    mock_flow.hass = mock_hass
```
this is not how we set up config flow tests, check the other integrations

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,98 @@
+
+
+@pytest.mark.asyncio
+async def test_user_step_invalid_auth():
```
those 3 tests can be one parametrized test. Also, in the end of the test remove the error and finish the flow successfully to show we can recover from errors

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,63 @@
+@pytest.fixture
+def switch(hass: HomeAssistant, switch_data):
+    """Return a RedgtechSwitch instance."""
+    coordinator = MagicMock()
+    coordinator.data = [switch_data]
+    coordinator.api.set_switch_state = AsyncMock()
+    coordinator.access_token = "test_access_token"
+    switch = RedgtechSwitch(coordinator, switch_data)
+    switch.hass = hass
+    switch.entity_id = switch_data.entity_id 
+    return switch
```
don't overwrite HA objects, rather mock the objects required to setup correctly

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,63 @@
+@pytest.mark.asyncio
+async def test_turn_on_switch(switch):
+    """Test turning on the switch."""
+    await switch.async_turn_on()
+    switch.coordinator.api.set_switch_state.assert_called_once_with("1234-5678", True, "test_access_token")
+    assert switch.device.state is STATE_ON
+
+@pytest.mark.asyncio
+async def test_turn_off_switch(switch):
+    """Test turning off the switch."""
+    switch.device.state = True
+    await switch.async_turn_off()
+    switch.coordinator.api.set_switch_state.assert_called_once_with("1234-5678", False, "test_access_token")
+    assert switch.device.state is STATE_OFF
```
is not how we test service calls, check other switch tests how it's done correctly

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+            self.device.state = STATE_ON if state else STATE_OFF
+            self.async_write_ha_state()
+        except RedgtechAuthError:
+            await self.coordinator.renew_token()
```
needs a `try` to result in `HomeAssistantError` on failure

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  entity-unavailable: done
+  integration-owner: done
+  log-when-unavailable: done
+  parallel-updates: done
```
todo

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+rules:
+  # Bronze
+  action-setup: done
```
exempt, no custom actions

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions: done
```
exempt, no custom actions

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+        """Initialize the switch."""
+        self.coordinator = coordinator
+        self.device = device
+        self._attr_unique_id = f"redgtech_{device.id}"
```
what is the device id? a `uuid`?

Jonhsady replied:
yes, uuid.

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,87 @@
+            identifiers={(DOMAIN, device.id)},
+            name=device.name,
+            manufacturer="Redgtech",
+            model="Switch Model",
```
this is not a valid model

Jonhsady replied:
What can you suggest I do?

Jonhsady replied:
model=device.model if hasattr(device, "model") else "Unknown Model"
resolveria?

zweckj replied:
ideally get the actual model from your API if you can. Otherwise don't set it

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  discovery-update-info:
+    status: exempt
+    comment: Discovery not possible
+  discovery:
+    status: exempt
+    comment: Discovery not possible
```
how do your devices connect to the cloud? Are they connected to WiFi?

Jonhsady replied:
yes, conneted to wifi

zweckj replied:
then this is todo, as we can probably discover it thorugh DHCP

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: done
```
todo

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: done
+  entity-device-class: done
```
todo

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  dynamic-devices: todo
+  entity-category: done
+  entity-device-class: done
+  entity-disabled-by-default: done
```
todo

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  icon-translations: done
+  reconfiguration-flow: todo
+  repair-issues: done
+  stale-devices: done
```
todo

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  stale-devices: done
+
+  # Platinum
+  async-dependency: done
+  inject-websession: done
+  strict-typing: done
```
```suggestion
  async-dependency: done
  inject-websession: todo
  strict-typing: todo
```

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  exception-translations: todo
+  icon-translations: done
+  reconfiguration-flow: todo
+  repair-issues: done
```
exempt, no repairs at the moment

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,64 @@
+  test-coverage: done
+  # Gold
+  devices: done
+  diagnostics: done
```
todo

Jonhsady committed:
adjustments

zweckj commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,27 @@
+      "user": {
+        "title": "User configuration",
+        "description": "Please enter your email address.",
+        "data": {
+          "email": {
+            "label": "Email",
+            "description": "Enter your email address for authentication."
+          },
+          "password": {
+            "label": "Password",
+            "description": "Enter your password for authentication."
+          }
```
this is not right. We use `data` and `data_description` objects and you can use common string referenes check https://github.com/zweckj/home-assistant-core/blob/dev/homeassistant/components/lamarzocco/strings.json

zweckj commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,27 @@
+          }
+        },
+        "errors": {
+          "invalid_auth": "Invalid authentication.",
+          "invalid_email": "Invalid email address.",
+          "invalid_password": "Invalid password.",
+          "email_required": "Email is required.",
+          "password_required": "Password is required."
```
missing the correct ones you need, you can use common strings here

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,83 @@
+            try:
+                await self.coordinator.renew_token()
+                await self.coordinator.api.set_switch_state(self.device.id, new_state, self.coordinator.access_token)
+                self.device.state = STATE_ON if new_state else STATE_OFF
+                self.async_write_ha_state()
```
move out of try

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,83 @@
+        if self.is_on:
+            await self.async_toggle()
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
```
```suggestion
```
you don't need it it's doing nothing extra

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+FAKE_TOKEN = "fake_token"
+
+
+@pytest.fixture
+def mock_redgtech_login():
+    """Mock the RedgtechAPI login method."""
+    with patch("homeassistant.components.redgtech.config_flow.RedgtechAPI.login", return_value=FAKE_TOKEN) as mock:
+        yield mock
+
+
+@pytest.fixture
+def mock_redgtech_invalid_auth():
+    """Mock the RedgtechAPI login method to raise an authentication error."""
+    with patch("homeassistant.components.redgtech.config_flow.RedgtechAPI.login", side_effect=RedgtechAuthError) as mock:
+        yield mock
+
+
+@pytest.fixture
+def mock_redgtech_cannot_connect():
+    """Mock the RedgtechAPI login method to raise a connection error."""
+    with patch("homeassistant.components.redgtech.config_flow.RedgtechAPI.login", side_effect=RedgtechConnectionError) as mock:
+        yield mock
+
```
Make a single fixture for `homeassistant.components.redgtech.config_flow.RedgtechAPI.login` which you can manipulate as needed

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+        CONF_PASSWORD: TEST_PASSWORD,
+    }
+
+    with patch(
+        "homeassistant.components.redgtech.config_flow.RedgtechAPI.login",
+        side_effect=side_effect,
```
use the fixutre

zweckj replied:
not resolved

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+        result = await mock_flow.async_step_user(user_input)
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"]["base"] == expected_error
```
after that remove the error and finish the test successfully to show we can recover from errors

zweckj replied:
you did not do that

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+        CONF_EMAIL: TEST_EMAIL,
+        CONF_PASSWORD: TEST_PASSWORD,
+    }
+
```
add a test for trying to add a duplicate entry

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,116 @@
+    return device
+
+
+@pytest.fixture
+def mock_coordinator(mock_device):
+    """Return a mocked RedgtechDataUpdateCoordinator."""
+    coordinator = AsyncMock()
+    coordinator.data = [mock_device]
+    coordinator.api = AsyncMock()
+    coordinator.api.set_switch_state = AsyncMock()
+    coordinator.access_token = "test_access_token"
+    return coordinator
```
I already told you last time: don't mock home assistant internals, instead mock the objects from your library you need to work

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,116 @@
+    """Set up the Redgtech switch for testing."""
+    config_entry.add_to_hass(hass)
+
+    with patch(
+        "homeassistant.components.redgtech.RedgtechDataUpdateCoordinator",
+        return_value=mock_coordinator,
```
same, don't

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,116 @@
+    return entity_id
+
+
+async def test_switch_initial_state(hass: HomeAssistant, setup_switch):
+    """Test the initial state of the switch."""
+    entity_id = setup_switch
+    state = hass.states.get(entity_id)
+    assert state is not None
```
use the snapshot_platform helper instead

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,116 @@
+
+
+@pytest.fixture
+async def setup_switch(hass: HomeAssistant, config_entry, mock_coordinator, mock_device):
```
please use more standard ways to setup your platform (check other components that use `snapshot_platform`)

zweckj replied:
this is also unresolved

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,98 @@
+        except RedgtechConnectionError as e:
+            raise UpdateFailed("Failed to connect to Redgtech API") from e
+        except HomeAssistantError as e:
+            raise UpdateFailed("Home Assistant-specific error while fetching data") from e
```
this solves except HomeAssistantError as e:
raise UpdateFailed(f"Error while fetching data: {e}") from e?

zweckj replied:
I know but you are making it a HomeAssistantError so it's originally something different. As I said, better handle them completely separately instead of doing that mix you currently have.

Jonhsady replied:
except RedgtechAuthError as e:
                raise ConfigEntryError("Authentication error while renewing access token") from e
            except RedgtechConnectionError as e:
                raise UpdateFailed("Failed to connect to Redgtech API") from e
            except Exception as e:
                raise UpdateFailed("Unexpected error during token renewal or data fetch") from e}

Jonhsady replied:
Thanks for the feedback, that makes sense. Since the HomeAssistantError is not thrown inside coordinator.py, I'll remove this except block completely and leave just the generic except Exception as a fallback.

Does this solve the issue or would you prefer me to handle a specific error instead?

Jonhsady committed:
adjustments

zweckj commented on homeassistant/components/redgtech/__init__.py:
```diff
@@ -0,0 +1,34 @@
+    _LOGGER.debug("Unloading Redgtech entry: %s", entry.entry_id)
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+__ALL__ = ["RedgtechApi"]
```
```suggestion
```

zweckj commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,53 @@
+        
+        errors: dict[str, str] = {}
+
+        if user_input:
```
```suggestion
        if user_input is not None:
```

zweckj commented on homeassistant/components/redgtech/const.py:
```diff
@@ -0,0 +1,2 @@
+DOMAIN = "redgtech"
+API_URL = "https://redgtech-dev.com"
```
```suggestion
```
unused I believe

Jonhsady replied:
remove, done!

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,99 @@
+@dataclass
+class RedgtechDevice:
+    """Representation of a Redgtech device."""
+    id: str
```
```suggestion
    unique_id: str
```
`id` is a built in function from python, let's not override that

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,99 @@
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(minutes=1),
```
make the update_interval a const please (in this file)

Jonhsady replied:
done

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,99 @@
+        """Login to the Redgtech API and return the access token."""
+        try:
+            access_token = await self.api.login(email, password)          
+        except Exception as e:
```
don't catch base exceptions outside of the config flow

Jonhsady replied:
done

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,99 @@
+            access_token = await self.api.login(email, password)          
+        except Exception as e:
+            raise ConfigEntryError("Unexpected error during login") from e
+        self.access_token = access_token
+        return access_token
```
it's strange that this sets `self.access_token` and then you return the value which you also use to override `self.access_token`. I think you can just make this return `None` and mark `access_token` private for the coordinator

Jonhsady replied:
async def login(self, email: str, password: str) -> None:
        """Login to the Redgtech API and store the access token."""
        try:
            self.access_token = await self.api.login(email, password)
        except RedgtechAuthError as e:
        
        
        
like this?

zweckj replied:
yes

zweckj commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,99 @@
+
+            data = await self.api.get_data(self.access_token)
+        except RedgtechAuthError:
+            _LOGGER.warning("Auth failed, trying to renew token")
```
```suggestion
            _LOGGER.debug("Auth failed, trying to renew token")
```
I don't think that's worth a warning

Jonhsady replied:
done

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,66 @@
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
+  entity-event-setup: done
```
```suggestion
  entity-event-setup: exempt
```
comment: no explicit subscriptions to events

zweckj commented on homeassistant/components/redgtech/quality_scale.yaml:
```diff
@@ -0,0 +1,66 @@
+  integration-owner: done
+  log-when-unavailable: done
+  parallel-updates: todo
+  reauthentication-flow: done
```
```suggestion
  reauthentication-flow: todo
```

Jonhsady replied:
done

zweckj commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,52 @@
+          "password": "[%key:common::config_flow::data_description::password%]"
+        }
+      },
+      "reauth_confirm": {
+        "description": "Re-authentication required. Please enter your password again.",
+        "data": {
+          "password": "[%key:common::config_flow::data::password%]"
+        },
+        "data_description": {
+          "password": "[%key:common::config_flow::data_description::password%]"
+        }
```
```suggestion
```
currently no reauth

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,88 @@
+                await self.coordinator.api.set_switch_state(
+                    self.device.id, new_state, self.coordinator.access_token
+                )
+                self.device.state = STATE_ON if new_state else STATE_OFF
+                self.async_write_ha_state()
```
```suggestion
```
not needed

Jonhsady replied:
When I remove this line the button activates, sends the correct command to activate, but then it returns as if it was not activated.

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,88 @@
+                raise HomeAssistantError("Failed to set switch state due to authentication error") from e
+        except RedgtechConnectionError as e:
+            raise HomeAssistantError("Connection error while setting switch state") from e
+        except Exception as e:
+            raise HomeAssistantError("Unexpected error while toggling switch state") from e
```
don't catch base exceptions

Jonhsady replied:
done

zweckj commented on homeassistant/components/redgtech/strings.json:
```diff
@@ -0,0 +1,52 @@
+    }
+  },
+  "entity": {
+    "switch": {
+      "device_switch": {
+        "name": "Device Switch"
+      }
```
```suggestion
```
unused

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,88 @@
+    coordinator = config_entry.runtime_data
+    async_add_entities(RedgtechSwitch(coordinator, device) for device in coordinator.data)
+
+class RedgtechSwitch(ToggleEntity):
```
```suggestion
class RedgtechSwitch(CoordinatorEntity[RedgtechDataUpdateCoordinator], SwitchEntity):
```
also make sure to call `super()` now in the constructor

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+@pytest.fixture
+def mock_redgtech_login():
+    """Mocka o método login da RedgtechAPI para retornar um token falso."""
```
english everywhere please

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+    ],
+)
+async def test_user_step_errors(hass: HomeAssistant, side_effect, expected_error):
+    """Testa tratamento de erros no passo do usuário."""
```
idem

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+    [
+        (RedgtechAuthError, "invalid_auth"),
+        (RedgtechConnectionError, "cannot_connect"),
+        (Exception("Erro genérico"), "unknown"),
```
idem

Jonhsady replied:
done

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+
+async def test_user_step_creates_entry(hass: HomeAssistant, mock_redgtech_login):
+    """Testa criação correta de entrada na configuração."""
```
for the rest of the comments as well

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+    result = await flow.async_step_user(user_input)
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

Jonhsady replied:
done

zweckj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+    user_input = {CONF_EMAIL: TEST_EMAIL, CONF_PASSWORD: TEST_PASSWORD}
+
+    # Captura a exceção AbortFlow
+    with pytest.raises(AbortFlow) as exc_info:
+        await flow.async_step_user(user_input)
```
instead of expecting the exception do ` assert result["type"] is FlowResultType.ABORT`
+ `reason`

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,128 @@
+"""Teste da plataforma switch Redgtech."""
```
idem

zweckj commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,128 @@
+    return device
+
+
+@pytest.fixture
+def mock_coordinator(mock_device):
+    """Cria um coordenador simulado."""
+    coordinator = AsyncMock()
+    coordinator.data = [mock_device]
+    coordinator.api = AsyncMock()
+    coordinator.api.set_switch_state = AsyncMock()
+    coordinator.access_token = "token_teste"
+    coordinator.email = "teste@exemplo.com"
+    coordinator.password = "senha123"
+    return coordinator
```
don't mock the coordinator, mock your API instead (with a fixture)

Jonhsady replied:
done

Jonhsady committed:
adjustments

zweckj reviewed (changes requested):
still a bunch of things not done from last review

Jonhsady committed:
adjustements

zweckj commented on homeassistant/components/redgtech/manifest.json:
```diff
@@ -0,0 +1,11 @@
+    "integration_type": "service",
+    "config_flow": true,
+    "quality_scale": "bronze",
+    "requirements": ["redgtech-api==0.1.25"]
```
library repo is still missing tagged releases, published through CICD to PyPi

Jonhsady replied:
https://pypi.org/project/redgtech-api

zweckj replied:
<img width="498" height="205" alt="Screenshot 2025-07-17 at 08 41 03" src="https://github.com/user-attachments/assets/7e49b7d6-62a6-4fef-8b68-8316b318b1a1" />

Jonhsady committed:
adjustments

Jonhsady committed:
adjustments

joostlek reviewed (dismissed):
Please use a proper development environment

Jonhsady committed:
adjustments

joostlek commented:
Do you now have a proper development environment? Mind adding one line to a file and then trying to commit it

Jonhsady committed:
test: verify development environment setup

joostlek commented:
Mypy.ini is okay, the other is not for this PR, but I would expect more issues and formatting things to come out of this

Jonhsady committed:
fix: Redgtech integration adjustments

Jonhsady committed:
adjustments

Jonhsady committed:
feat: Fix Redgtech integration state synchronization and linting issues

- Fix real-time state synchronization in UI
- Add proper state management with async_write_ha_state()
- Fix import sorting and whitespace issues
- All tests passing (15/15)
- Linting checks passed

Jonhsady committed:
feat: Fix Redgtech integration state synchronization and linting issues

- Fix real-time state synchronization in UI
- Add proper state management with async_write_ha_state()
- Fix import sorting and whitespace issues
- All tests passing (15/15)
- Linting checks passed

Jonhsady committed:
fix: Update redgtech-api to version 0.1.30 to resolve hassfest requirements error

Jonhsady committed:
fix: Update redgtech-api to version 0.1.32 with correct GitHub URL and synchronized releases

emontnemery commented:
@Jonhsady There are still multiple issues:
- CI is still failing.
- It looks like you might have messed up your library repo, there's now a single visible commit https://github.com/redgtech-automacao/redgtech-python-api/commit/cb0ea68662668a930e843089f96a0d9f0f088ae1 across the 3 branches.
- The library still [doesn't meet the documented requirements](https://developers.home-assistant.io/docs/api_lib_index?_highlight=libr#basic-library-requirements), these rules don't seem to be met:
  - The library versions published on PyPi should correspond to tagged releases in a public online repository.
  - The publishing on PyPi must be automated.

I think you might have tried to fix that by adding your local builds to the library repo, but that's not what the requirement is about, the requirement is about relying on github CI for publishing to PyPi instead of you doing it locally from the command line. Here's an example of how it can be implemented: https://github.com/emontnemery/py-dormakaba-dkey/blob/main/.github/workflows/pythonpublish.yml

Jonhsady committed:
fix: update redgtech-api to version 0.1.33

Jonhsady committed:
fix: update redgtech-api to version 0.1.34

Jonhsady committed:
fix: update redgtech-api to version 0.1.34

Jonhsady committed:
Fix pylint type hints warnings

- Change AddEntitiesCallback to AddConfigEntryEntitiesCallback in switch.py
- Add return type annotation -> None to __init__ in coordinator.py
- Add FrozenDateTimeFactory type annotation for freezer parameter in tests
- Fix W7431 and W7432 pylint warnings

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+
+from .const import DOMAIN
+
+UPDATE_INTERVAL = timedelta(seconds=10)
```
the api supports websocket can i change iot to cloud_push?

Jonhsady committed:
adjustments

Jonhsady committed:
adjustments

emontnemery commented:
Set to draft, please change back to "ready for review" once the type annotations in the library are fixed.

Jonhsady commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,106 @@
+
+            self._async_abort_entries_match({CONF_EMAIL: user_input[CONF_EMAIL]})
+
+            api: Any = RedgtechAPI()  # type: ignore[no-untyped-call]
```
this is ok?

class RedgtechAPI:
    def __init__(self, token: str | None = None) -> None:
        self._token: str | None = token
        self._session: aiohttp.ClientSession | None = None

    async def _get_session(self) -> aiohttp.ClientSession:
        """Get or create aiohttp session."""
        if self._session is None:
            self._session = aiohttp.ClientSession()
        return self._session

    async def close(self) -> None:
        if self._session:
            await self._session.close()
            self._session = None
            
            I included str | None in the token because it resolves another mypy error. I included str | None in the token because it resolves another mypy error.

mypy complained in switch.py: “Argument 3 to set_switch_state ... has incompatible type 'str | None'; expected 'str'.”

Jonhsady committed:
adjustments api

Jonhsady committed:
reauth in subsequent pr

Jonhsady committed:
adjuste stringJson for tests

Jonhsady committed:
Update snapshots and clean up error messages

- Update entity registry snapshots for compatibility
- Remove duplicate error messages from strings.json
- All tests passing

emontnemery commented:
The failing tests seem to be mostly located to `strings.json`

Do you have some local changes which you've forgotten to commit and push?
Have you run `python3 -m script.translations develop --all` locally?

Jonhsady committed:
translations for test

Jonhsady commented:
> Os testes com falha parecem estar localizados principalmente em`strings.json`
> 
> Você tem alguma alteração local que esqueceu de enviar e enviar? Você executou `python3 -m script.translations develop --all`localmente?


After running python3 -m script.translations develop --all I received the translation error, I made the correction and the tests passed

emontnemery commented:
Your local branch is also very outdated.

I'd suggest that you do this (assuming you've added `https://github.com/home-assistant/core` as `upstream` and your fork as `origin`):
```
git pull upstream
git merge upstream/dev
git push origin
```

Jonhsady committed:
Merge remote-tracking branch 'upstream/dev' into dev

Jonhsady committed:
Update snapshots after upstream merge

Jonhsady commented:
> Your local branch is also very outdated.
> 
> I'd suggest that you do this (assuming you've added `https://github.com/home-assistant/core` as `upstream` and your fork as `origin`):
> 
> ```
> git pull upstream
> git merge upstream/dev
> git push origin
> ```

<img width="1860" height="494" alt="image" src="https://github.com/user-attachments/assets/531193f7-f457-40b0-9ea4-64ccd528f4d2" />

emontnemery commented on homeassistant/components/redgtech/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+                    vol.Required(CONF_EMAIL): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
```
Wrap the schema in `add_suggested_values_to_schema` so the user doesn't need to re-enter email and password if they made a mistake.
```suggestion
            data_schema=self.add_suggested_values_to_schema(
                vol.Schema(
                    {
                        vol.Required(CONF_EMAIL): str,
                        vol.Required(CONF_PASSWORD): str,
                    }
                )
                user_input,
            ),
```

emontnemery commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,120 @@
+        try:
+            await self.ensure_token()
+
+            data = await self.api.get_data(self.access_token or "")
```
There's no need to fall back to "" here
```suggestion
            data = await self.api.get_data(self.access_token)
```

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,120 @@
+                    "endpointId": item["endpointId"],
+                    "friendlyName": item["friendlyName"],
+                    "value": item["value"],
+                    "type": item.get("type", "switch"),
```
I don't always return the field but I'm open to suggestions, I want to adapt to you.

Jonhsady committed:
add placeholder

Jonhsady committed:
add type dinamic

emontnemery commented on homeassistant/components/redgtech/device.py:
```diff
@@ -0,0 +1,32 @@
+from typing import Any
+
+
+@dataclass
```
This is not a good fit for a dataclass IMO considering the custom `__init__` method
I'd suggest to remove the custom `__init__` and let the coordinator pass in the values directly instead.

emontnemery commented on homeassistant/components/redgtech/device.py:
```diff
@@ -0,0 +1,32 @@
+        self.type = data.get("type", "switch")
+        self.state = data.get("value", False)
+        self.model = data.get("model")
+        self.version = data.get("version")
```
Just remove this

emontnemery commented on homeassistant/components/redgtech/device.py:
```diff
@@ -0,0 +1,32 @@
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this device."""
+        return self.id
```
Why don't you just change the name of the `id` attribute to `unique_id`?

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,111 @@
+
+        for item in data["boards"]:
+            display_categories = item["displayCategories"]
+            device_type = display_categories[0].lower()
```
Yes, the API can return multiple categories (for example, ["LIGHT", "SWITCH"]). I'll adjust the code to handle more categories.

emontnemery replied:
I'm not sure exactly what you're asking, so bear with me.

If you're asking about the integration supporting several kind of entities:

The idea is that in a PR introducing a new integration, you add support for only one type of entity. In this PR you've chosen to add support for `switch` entities.

Once this PR is approved and merged, you're welcome to open new PRs with additional features:
- New kind of entities (for example light)
- Better error handling (for example a reauth flow)

If you're asking if an endpoint should be added as both a light and a switch:
- If the endpoint which is `["LIGHT", "SWITCH"]` is the same physical thing, meaning something like "light" supports adjusting the brightness and "switch" supports turning on and off, you should only add the endpoint as a light with brightness support (in a follow-up PR)

Jonhsady committed:
adjustments in device

Jonhsady committed:
adjustments append device

Jonhsady committed:
add newline at end of file

Jonhsady committed:
adjustmentes append device

Jonhsady committed:
Filter Redgtech switch devices to exclude light-capable endpoints

Jonhsady commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,111 @@
+                type="switch",
+            )
+            _LOGGER.debug("Processing device: %s", device)
+            devices.append(device)
```
@emontnemery  I await any possible feedback.

Jonhsady replied:
@emontnemery Should I wait, or has something unexpected happened?

emontnemery replied:
Sorry about the lack of updates from my side, I was very busy the last couple of months.

emontnemery commented:
(I closed + opened the PR to re-run CI)

emontnemery commented:
@Jonhsady Please sort the JSON file

Jonhsady committed:
Sort string.json keys alphabetically

emontnemery commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,309 @@
+            blocking=True,
+        )
+
+    # Test completed successfully
```
```suggestion
```

emontnemery commented on tests/components/redgtech/test_switch.py:
```diff
@@ -0,0 +1,309 @@
+                    "friendlyName": "Kitchen Switch",
+                    "value": True,
+                    "displayCategories": ["SWITCH"],
+                },
```
Could you add a third board here which has `"displayCategories": ["LIGHT", "SWITCH"],` to check it's not added as a switch?

emontnemery commented on homeassistant/components/redgtech/coordinator.py:
```diff
@@ -0,0 +1,111 @@
+            _LOGGER.debug("No access token, logging in")
+            self.access_token = await self.login(self.email, self.password)
+        else:
+            _LOGGER.debug("Using existing access token")
```
This method is not doing what the docstring says it does; the token is not automatically renewed.
If it's not possible to detect if the token needs renewal without calling the API, maybe do a helper like this which can retry any API call if there's an auth error:

```py
    from collections.abc import Coroutine
    from typing import Any

    async def call_api_with_valid_token[T](self, api_call: Coroutine[Any, Any, T]) -> T:
        """Make an API call with a valid token.

        Ensure we have a valid access token, renewing it if necessary.
        """
        if not self.access_token:
            _LOGGER.debug("No access token, logging in")
            self.access_token = await self.login(self.email, self.password)
        else:
            _LOGGER.debug("Using existing access token")
        try:
            return await api_call
        except RedgtechAuthError:
            _LOGGER.debug("Auth failed, trying to renew token")
            await self.renew_token(
                self.config_entry.data[CONF_EMAIL],
                self.config_entry.data[CONF_PASSWORD],
            )
            return await api_call
```

Then you call it like this from the coordinator:

```py
    async def _async_update_data(self) -> list[RedgtechDevice]:
        """Fetch data from the API on demand."""
        _LOGGER.debug("Fetching data from Redgtech API on demand")
        try:
             data = await self.call_api_with_valid_token(
                self.api.get_data(self.access_token)
            )
        except RedgtechAuthError as e:
            raise UpdateFailed("Authentication failed") from e
        except RedgtechConnectionError as e:
            raise UpdateFailed("Failed to connect to Redgtech API") from e

        devices: list[RedgtechDevice] = []

        ...
```

And like this from the switch:

```py
    async def _set_state(self, new_state: bool) -> None:
        """Set state of the switch."""
        try:
            await self.coordinator.call_api_with_valid_token(
                self.coordinator.api.set_switch_state(
                    self.device.unique_id, new_state, self.coordinator.access_token
                )
            )
            await self.coordinator.async_refresh()
        ...
```

emontnemery reviewed (changes requested):
A couple of very small remarks on the tests, and one suggestion to improve the token renewal logic.

zweckj commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,105 @@
+            return False
+
+        for device in self.coordinator.data:
+            if device.unique_id == self.device.unique_id:
```
instead of looping through the list every time we want to read the state, updating the device during `_async_set_updated_data` makes more sense imo. Also, a smarter data structure for the coordinator's data will help us avoid loops

zweckj commented on tests/components/redgtech/test_switch.py:
please use a common api mock from conftest for both platform and config flow tests

zweckj commented on tests/components/redgtech/test_config_flow.py:
let your fixture return an instance so you don't have to set it up every time. See `tedee` for an example how you can do that for both platform and config flow at once

mik-laj commented on tests/components/redgtech/test_config_flow.py:
```diff
@@ -0,0 +1,107 @@
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"]["base"] == expected_error
+    mock_api.login.assert_called_once_with(TEST_EMAIL, TEST_PASSWORD)
```
Can these tests confirm that the flow can recover from an error to ensure that the user can complete the flow even if something goes wrong?

It is needed according to rules: 

> This means that we want to have 100% test coverage for the config flow. In those tests, we require verification that the flow is able to recover from an error to confirm that the user is able to finish the flow even if something goes wrong.

https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/config-flow-test-coverage

Jonhsady commented on homeassistant/components/redgtech/switch.py:
```diff
@@ -0,0 +1,105 @@
+    def is_on(self) -> bool:
+        """Return true if the switch is on."""
+        if not self.coordinator.data:
+            return False
```
ok

Jonhsady committed:
redgtech: Refactor API calls, optimize data access and improve tests

- Sort JSON keys alphabetically in strings.json
- Add automatic token renewal on auth errors in coordinator
- Optimize coordinator data structure from list to dict for O(1) access
- Move RedgtechDevice dataclass to coordinator.py
- Refactor test fixtures to use common mock instance
- Add error recovery tests for config flow

Jonhsady committed:
adjustments test_config_flow  and quality_scale

zweckj reviewed (changes requested):
I have a last set of nits I missed earlier, but I think then we're good.
Please also make sure to always run the pre-commit checks in your dev environment, because CI is currently failing as there was an extra space added from my last suggestion.

Jonhsady committed:
redgtech: Refactor API calls, optimize data access and improve code quality

- Add automatic token renewal on auth errors
- Optimize coordinator data structure from list to dict
- Move RedgtechDevice to coordinator.py
- Remove redundant logging
- Use SOURCE_USER constant in tests
- Update quality_scale.yaml
- Improve test fixtures and add error recovery tests

Jonhsady committed:
Correcting spacing

zweckj reviewed (dismissed):
Tests are failing. Rebase your branch and regenerate snapshots

emontnemery reviewed (approved):
This is good to go now, except for the snapshots.

Since the tick-box to accept changes from maintainers is disabled, we can merge it as is and fix the snapshots ourselves in a follow-up.
